packages/simple-cache-provider/src/SimpleCacheProvider.js

```diff
@@ -0,0 +1,201 @@
+          'instead received: %s. Strings and numbers are not permitted as ' +
+          'resource types.',
+        resourceType,
+      );
+    }
+
+    let recordCache = resourceCache.get(resourceType);
+    if (recordCache !== undefined) {
+      const record = recordCache.get(key);
+      if (record !== undefined) {
+        return record;
+      }
+    } else {
+      recordCache = new Map();
+      resourceCache.set(resourceType, recordCache);
+    }
+
+    const record = {
+      status: Empty,
+      suspender: null,
+      value: null,
+      error: null,
+    };
+    recordCache.set(key, record);
+    return record;
+  }
+
+  function load<K, V>(record: Record<V>, key: K, miss: K => Promise<V>) {
+    const suspender = miss(key);
+    record.suspender = suspender;
+    suspender.then(
+      value => {
+        // Resource loaded successfully.
+        record.suspender = null;
+        record.status = Resolved;
+        record.value = value;
+      },
+      error => {
+        // Resource failed to load. Stash the error for later so we can throw it
+        // the next time it's requested.
+        record.suspender = null;
+        record.status = Rejected;
+        record.error = error;
+      },
+    );
+    return suspender;
+  }
+
+  const cache: Cache = {
+    invalidate() {
+      invalidator();
+    },
+    preload<K, V>(resourceType: any, key: K, miss: K => Promise<V>): void {
+      const record: Record<V> = getRecord(resourceType, key);
+      const suspender = record.suspender;
+      if (suspender !== null) {
+        // There's already a pending request.
+        return;
+      }
+      switch (record.status) {
+        case Empty:
+          // Warm the cache.
+          load(record, key, miss);
+          return;
+        case Resolved:
+        case Rejected:
+          // The resource is already in the cache.
+          return;
+      }
+    },
+    read<K, V>(resourceType: any, key: K, miss: K => Promise<V>): V {
+      const record: Record<V> = getRecord(resourceType, key);
+      const suspender = record.suspender;
+      if (suspender !== null) {
+        // There's already a pending request.
+        throw suspender;
+      }
+      switch (record.status) {
+        case Empty:
+          // Load the requested resource.
+          throw load(record, key, miss);
+        case Resolved:
+          return ((record.value: any): V);
+        case Rejected:
+        default:
+          // The requested resource previously failed loading.
+          // TODO: When should we try again?
+          throw record.error;
+      }
+    },
+  };
+
+  if (__DEV__) {
+    cache.$$typeof = CACHE_TYPE;
+  }
+  return cache;
+}
+
+type ResourceReader<K, V> = (Cache, K) => V;
+
+type Resource<K, V> = ResourceReader<K, V> & {
+  preload(cache: Cache, key: K): void,
+};
+
+export function createResource<K, V, H>(
+  loadResource: K => Promise<V>,
+  // $FlowFixMe: Flow complains that K !== H.
```

Comment:
Flow is right.

The return value of the hash should be limited to `string | number | boolean | undefined | null` (i.e. a primitive).

So therefore, if K is not a primitive, it is invalid to not pass a hash function. That's not modeled right now.

Reply:
You could possibly model this as overloaded functions.

```js
type primitive = string | number | boolean | void | null;
declare createResource<V>(loadResource: primitive => Promise<V>, hash?: primitive => primitive : V;
declare createResource<V>(loadResource: object => Promise<V>, hash: object => primitive) : V;
declare createResource<V>(loadResource: function => Promise<V>, hash: function => primitive) : V;
```

Reply:
I should use a sum type for Record, too.